t/balancer-stream.t

```diff
@@ -0,0 +1,298 @@
+--- stream_config
+    lua_package_path "$TEST_NGINX_CWD/lib/?.lua;;";
+
+    upstream backend {
+        server 0.0.0.1:80;
+        balancer_by_lua_block {
+            print("hello from balancer by lua!")
+            local b = require "ngx.balancer.stream"
+            assert(b.set_current_peer("127.0.0.3", 12345))
+        }
+    }
+--- stream_server_config
+    proxy_pass backend;
+--- error_log eval
+[
+'[lua] balancer_by_lua:2: hello from balancer by lua! while connecting to upstream,',
+qr{connect\(\) to failed .*?, upstream: "127\.0\.0\.3:12345"},
+]
+--- no_error_log
+[warn]
+
+
+
+=== TEST 2: set current peer & next upstream (3 tries)
+--- skip_nginx: 4: < 1.7.5
+--- stream_config
+    lua_package_path "$TEST_NGINX_CWD/lib/?.lua;;";
+
+    proxy_next_upstream_tries 10;
+
+    upstream backend {
+        server 0.0.0.1:80;
+        balancer_by_lua_block {
+            print("hello from balancer by lua!")
+            local b = require "ngx.balancer.stream"
+            if not ngx.ctx.tries then
+                ngx.ctx.tries = 0
+            end
+
+            if ngx.ctx.tries < 2 then
+                local ok, err = b.set_more_tries(1)
+                if not ok then
+                    return error("failed to set more tries: ", err)
+                elseif err then
+                    ngx.log(ngx.WARN, "set more tries: ", err)
+                end
+            end
+            ngx.ctx.tries = ngx.ctx.tries + 1
+            assert(b.set_current_peer("127.0.0.3", 12345))
+        }
+    }
+--- stream_server_config
+    proxy_pass backend;
+--- grep_error_log eval: qr{connect\(\) to failed .*, upstream: ".*?"}
+--- grep_error_log_out eval
+qr#^(?:connect\(\) to failed .*?, upstream: "127.0.0.3:12345"\n){3}$#
+--- no_error_log
+[warn]
+
+
+
+=== TEST 3: set current peer & next upstream (no retries)
+--- skip_nginx: 4: < 1.7.5
+--- stream_config
+    lua_package_path "$TEST_NGINX_CWD/lib/?.lua;;";
+
+
+    upstream backend {
+        server 0.0.0.1:80;
+        balancer_by_lua_block {
+            print("hello from balancer by lua!")
+            local b = require "ngx.balancer.stream"
+            if not ngx.ctx.tries then
+                ngx.ctx.tries = 0
+            end
+
+            ngx.ctx.tries = ngx.ctx.tries + 1
+            assert(b.set_current_peer("127.0.0.3", 12345))
+        }
+    }
+--- stream_server_config
+    proxy_pass backend;
+--- grep_error_log eval: qr{connect\(\) failed .*, upstream: ".*?"}
+--- grep_error_log_out eval
+qr#^(?:connect\(\) to failed .*?, upstream: "127.0.0.3:12345"\n){1}$#
+--- no_error_log
+[warn]
+
+
+
+=== TEST 4: set current peer & next upstream (3 tries exceeding the limit)
+--- skip_nginx: 4: < 1.7.5
+--- stream_config
+    lua_package_path "$TEST_NGINX_CWD/lib/?.lua;;";
+
+    proxy_next_upstream_tries 2;
+
+    upstream backend {
+        server 0.0.0.1:80;
+        balancer_by_lua_block {
+            local b = require "ngx.balancer.stream"
```

Comment:
We should use `ngx.balancer` as the API while keeping the implementation in `ngx.balancer.stream`. So the user require `ngx.balancer` instead of `ngx.balancer.stream`. This encourages Lua code reuse on the user Lua land.